lockers/k8s_locker/k8s_registration.go

```diff
@@ -22,59 +24,124 @@ func (k *k8sLocker) Deregister(s string) error {
 			if err != nil {
-				return err
+				k.logger.Printf("watch ended with error: %s", err)
+				time.Sleep(k.Cfg.RetryTimer)
+			} else if k.Cfg.Debug {
+				k.logger.Print("watch timed out")
 			}
-			sChan <- services
-			time.Sleep(watchTimeout)
 		}
 	}
 }
 
-func (k *k8sLocker) GetServices(ctx context.Context, serviceName string, tags []string) ([]*lockers.Service, error) {
-	ksrv, err := k.clientset.CoreV1().Services(k.Cfg.Namespace).Get(ctx, serviceName, metav1.GetOptions{})
-	if err != nil {
-		return nil, err
+func (k *k8sLocker) watch(ctx context.Context, serviceName string, tags []string, sChan chan<- []*lockers.Service, watchTimeout time.Duration, resourceVersion string) (string, error) {
+	timeoutSeconds := int64(watchTimeout.Seconds())
+	listopts := metav1.ListOptions{
+		FieldSelector:        fields.OneTermEqualSelector(metav1.ObjectNameField, serviceName).String(),
+		ResourceVersion:      resourceVersion,
+		TimeoutSeconds:       &timeoutSeconds,
 	}
-	if len(ksrv.Spec.Ports) == 0 {
-		return nil, fmt.Errorf("missing ports in service %s", serviceName)
+	if k.Cfg.Debug {
+		if resourceVersion == "" {
+			k.logger.Print("starting watch beginning with unspecified resource version")
+		} else {
+			k.logger.Printf("starting watch beginning with resource version %s", resourceVersion)
+		}
 	}
-	portNum := ksrv.Spec.Ports[0].Port
-
-	set := labels.Set(ksrv.Spec.Selector)
-	pods, err := k.clientset.CoreV1().Pods(k.Cfg.Namespace).List(ctx, metav1.ListOptions{LabelSelector: set.AsSelector().String()})
+	watched, err := k.clientset.CoreV1().Endpoints(k.Cfg.Namespace).Watch(ctx, listopts)
 	if err != nil {
-		return nil, err
+		return "", err
 	}
+	defer watched.Stop()
+
+	watchChan := watched.ResultChan()
 
-	services := make([]*lockers.Service, 0, len(pods.Items))
-	for _, pod := range pods.Items {
-		if pod.Status.PodIP == "" {
-			continue
+	for {
+		event := <- watchChan
+		switch event.Type {
+		case watch.Modified, watch.Added:
+			endpoints, ok := event.Object.(*corev1.Endpoints)
+			if !ok {
+				// this ought not to happen, but we should probably
+				// start from scratch next time in case it does
+				return "", fmt.Errorf("error converting watch result to an endpoint")
+			}
+			resourceVersion = endpoints.ResourceVersion
+			if k.Cfg.Debug {
+				k.logger.Printf("received watch event %s for resource version %s", event.Type, resourceVersion)
+			}
+			svcs, err := parseEndpoint(endpoints)
+			if err != nil {
+				return "", err
+			}
+			sChan <- svcs
+		case "":
+			// reached the timeout. return the version we last saw so
+			// we can resume watching
+			return resourceVersion, nil
+		default:
+			// something else happened, including maybe the object we
+			// were watching being deleted. we'll need to start the
+			// next watch from scratch, so don't return the resource
+			// version
+			return "", fmt.Errorf("unexpected watch event: %s", event.Type)
```

Comment:
It would be good to include a `select` statement inside the `for{}` loop. It would selects on `<-watchChan` as well as `<-ctx.Done()`